src/ray/util/tests/pipe_logger_test.cc

```diff
@@ -185,6 +183,167 @@ TEST(PipeLoggerTestWithTee, RotatedRedirectionWithTee) {
   EXPECT_EQ(CompleteReadFile(log_file_path2), kLogLine1);
 }
 
+// Testing senario: log to stdout and file; check whether these two sinks generate
+// expected output.
+TEST(PipeLoggerCompatTest, CompatibilityTest) {
+  // Testing-1: No newliner in the middle nor at the end.
+  {
+    constexpr std::string_view kContent = "hello";
+    const std::string test_file_path = absl::StrFormat("%s.out", GenerateUUIDV4());
+
+    StreamRedirectionOption logging_option{};
+    logging_option.file_path = test_file_path;
+    logging_option.tee_to_stdout = true;
+
+    testing::internal::CaptureStdout();
+    auto stream_redirection_handle = CreateRedirectionFileHandle(logging_option);
+    stream_redirection_handle.CompleteWrite(kContent.data(), kContent.length());
+    stream_redirection_handle.Close();
+
+    const std::string stdout_content = testing::internal::GetCapturedStdout();
+    EXPECT_EQ(stdout_content, kContent);
+    EXPECT_EQ(CompleteReadFile(test_file_path), "hello\n");
+
+    EXPECT_TRUE(std::filesystem::remove(test_file_path));
+  }
+
+  // Testing-2: Newliner at the end.
+  {
+    constexpr std::string_view kContent = "hello\n";
+    const std::string test_file_path = absl::StrFormat("%s.out", GenerateUUIDV4());
+
+    StreamRedirectionOption logging_option{};
+    logging_option.file_path = test_file_path;
+    logging_option.tee_to_stdout = true;
+
+    testing::internal::CaptureStdout();
+    auto stream_redirection_handle = CreateRedirectionFileHandle(logging_option);
+    stream_redirection_handle.CompleteWrite(kContent.data(), kContent.length());
+    stream_redirection_handle.Close();
+
+    const std::string stdout_content = testing::internal::GetCapturedStdout();
+    EXPECT_EQ(stdout_content, kContent);
+    EXPECT_EQ(CompleteReadFile(test_file_path), "hello\n");
+
+    EXPECT_TRUE(std::filesystem::remove(test_file_path));
+  }
+
+  // Testing-3: Newliner in the middle.
+  {
+    constexpr std::string_view kContent = "hello\nworld";
+    const std::string test_file_path = absl::StrFormat("%s.out", GenerateUUIDV4());
+
+    StreamRedirectionOption logging_option{};
+    logging_option.file_path = test_file_path;
+    logging_option.tee_to_stdout = true;
+
+    testing::internal::CaptureStdout();
+    auto stream_redirection_handle = CreateRedirectionFileHandle(logging_option);
+    stream_redirection_handle.CompleteWrite(kContent.data(), kContent.length());
+    stream_redirection_handle.Close();
+
+    const std::string stdout_content = testing::internal::GetCapturedStdout();
+    EXPECT_EQ(stdout_content, kContent);
+    // Pipe logger doesn't know about the boundary for a whole user's content, so extra
+    // newliner is added in between.
```

Comment:
The reason is that spd logger always add a newline to the end and we cannot disable it.

Reply:
fixed